plugins/ccgx/fu-ccgx-dmc-firmware.c

```diff
@@ -0,0 +1,332 @@
+	return &self->fwct_info;
+}
+
+guint32
+fu_ccgx_dmc_firmware_get_fw_data_size (FuCcgxDmcFirmware *self)
+{
+	g_return_val_if_fail (FU_IS_CCGX_DMC_FIRMWARE (self), 0);
+	return self->fw_data_size;
+}
+
+static void
+fu_ccgx_dmc_firmware_to_string (FuFirmware *firmware, guint idt, GString *str)
+{
+	FuCcgxDmcFirmware *self = FU_CCGX_DMC_FIRMWARE (firmware);
+	fu_common_string_append_kx (str, idt, "FwDataSize", self->fw_data_size);
+	fu_common_string_append_ku (str, idt, "ImageRecords", self->image_records->len);
+	fu_common_string_append_kv (str, idt, "Version",
+				    fu_ccgx_dmc_version_to_string (self->fwct_info.composite_version));
+}
+
+static gboolean
+fu_ccgx_dmc_firmware_parse_segment (FuFirmware *firmware,
+				const guint8 *fw_buff,
+				gsize fw_buff_size,
+				FuCcgxDmcFirmwareImageRecord *img_rcd,
+				gsize* segment_info_offset,
+				GError **error)
+{
+	FuCcgxDmcFirmware *self = FU_CCGX_DMC_FIRMWARE (firmware);
+	gsize fw_buffer_offset = 0;
+	gsize row_data_offset = 0;
+	gsize digestlen = DMC_HASH_SIZE;
+	guint8 hash[DMC_HASH_SIZE];
+	g_autoptr(GChecksum) csum = g_checksum_new (G_CHECKSUM_SHA256);
+
+	/* set row data offset in current image */
+	row_data_offset = self->row_data_offset_start + img_rcd->info_header.img_offset;
+	/* create segment record array in image record */
+	img_rcd->seg_records = g_ptr_array_new_with_free_func ((GFreeFunc) fu_ccgx_dmc_firmware_segment_record_free);
+	/* parse segment in image */
+	for (guint32 seg_num = 0; seg_num < img_rcd->info_header.num_img_segments; seg_num++) {
+		guint16 actual_row_size = 0;
+		g_autoptr(FuCcgxDmcFirmwareSegmentRecord) seg_rcd = NULL;
+		g_autofree guint8 *row_buff = NULL;
+		/* alloc segment record */
+		seg_rcd = g_new0 (FuCcgxDmcFirmwareSegmentRecord, 1);
+		/* set segment info offset */
+		fw_buffer_offset = *segment_info_offset;
+		/* read segment info  */
+		if (!fu_memcpy_safe ((guint8 *)&seg_rcd->info_header, sizeof(seg_rcd->info_header), 0x0, /* dst */
+			     fw_buff, fw_buff_size, fw_buffer_offset, sizeof(seg_rcd->info_header), error)) /* src */
+			return FALSE;
+		/* calculate actual row size */
+		actual_row_size = img_rcd->info_header.row_size * 64;
+		/* alloc row buffer */
+		row_buff = g_malloc0 (actual_row_size);
+		/* create data record array in segment record */
+		seg_rcd->data_records = g_ptr_array_new_with_free_func ((GFreeFunc) fu_ccgx_dmc_firmware_data_record_free);
+		/* read row data in segment */
+		for (int row = 0; row < seg_rcd->info_header.num_rows; row++) {
+			g_autoptr(FuCcgxDmcFirmwareDataRecord) data_rcd = NULL;
+			/* alloc data record */
+			data_rcd = g_new0 (FuCcgxDmcFirmwareDataRecord, 1);
+			/* set row data offset */
+			fw_buffer_offset = row_data_offset;
+			/* read row data */
+			if (!fu_memcpy_safe (row_buff, actual_row_size, 0x0, /* dst */
+			     fw_buff, fw_buff_size, fw_buffer_offset, actual_row_size, error)) /* src */
+				return FALSE;
+			/* update hash */
```

Comment:
this whole section is a bit "dense" -- perhaps some newlines there?

Reply:
Added new line